scale-out-relay/src/main/scala/com/normation/plugins/scaleoutrelay/ScaleOutRelayService.scala

```diff
@@ -0,0 +1,264 @@
+  , woDirectiveRepository         : WoDirectiveRepository
+  , woRuleRepository              : WoRuleRepository
+  , uuidGen                       : StringUuidGenerator
+  , policyServerManagementService : PolicyServerManagementService
+  , actionLogger                  : EventLogRepository
+) {
+  val SYSTEM_GROUPS = "SystemGroups"
+  val DISTRIBUTE_POLICY = "distributePolicy"
+  val COMMON = "common"
+  val logger = NamedZioLogger("plugin.scaleoutrelay")
+
+  def promoteNodeToRelay(uuid: NodeId, actor: EventActor, reason:Option[String]): ZIO[Any, RudderError, NodeInfo] = {
+    val modId = ModificationId(uuidGen.newUuid)
+    for {
+      nodeInfos  <- getNodeToPromote(uuid)
+      targetedNode <-
+                    if (!nodeInfos.isPolicyServer) {
+                      val updatedNode = NodeToPolicyServer(nodeInfos)
+                      for {
+                         _        <- woLDAPNodeRepository.deleteNode(nodeInfos.node, modId, actor, reason)
+                                       .chainError(s"Remove ${nodeInfos.node.id.value} to update it to policy server failed")
+                         newRelay <- createRelayFromNode(updatedNode, modId, actor, reason)
+                      } yield {
+                        newRelay
+                      }
+                    } else {
+                      nodeInfos.succeed
+                    }
+    } yield {
+      targetedNode
+    }
+  }
+
+  private[scaleoutrelay] def createRelayFromNode(nodeInf: NodeInfo, modId: ModificationId, actor: EventActor, reason:Option[String]) = {
+    val directDistribPolicy     = createDirectiveDistributePolicy(nodeInf.node.id)
+    val ruleTarget              = createPolicyServer(nodeInf.node.id)
+    val nodeGroup               = createNodeGroup(nodeInf.node.id)
+    val ruleDistribPolicy       = createRuleDistributePolicy(nodeInf.node.id)
+    val ruleSetup               = createRuleSetup(nodeInf.node.id)
+    val categoryId              = NodeGroupCategoryId(SYSTEM_GROUPS)
+    val activeTechniqueId       = ActiveTechniqueId(DISTRIBUTE_POLICY)
+    val activeTechniqueIdCommon = ActiveTechniqueId(COMMON)
+
+    val promote = for {
+      commonDirective <- createCommonDirective(nodeInf).toIO
+
+      _ <- woLDAPNodeRepository.createNode(nodeInf.node, modId, actor, reason)
+      _ <- woLDAPNodeGroupRepository.createPolicyServerTarget(ruleTarget, modId, actor, reason)
+      _ <- woLDAPNodeGroupRepository.create(nodeGroup, categoryId, modId, actor, reason)
+      _ <- woDirectiveRepository.saveSystemDirective(activeTechniqueId, directDistribPolicy, modId, actor, reason)
+      _ <- woDirectiveRepository.saveSystemDirective(activeTechniqueIdCommon, commonDirective, modId, actor, reason)
+      _ <- woRuleRepository.create(ruleSetup, modId, actor, reason)
+      _ <- woRuleRepository.create(ruleDistribPolicy, modId, actor, reason)
+      _ <- actionLogger.savePromoteToRelay(modId,actor,nodeInf, reason)
```

Comment:
it will need a new PR in rudder core

Reply:
https://github.com/Normation/rudder/pull/2851